File: Solutions.cs/Simulation/0950. Reveal Cards In Increasing Order.cs
namespace Problem950
{

  /// <summary>
  /// 950. Reveal Cards In Increasing Order
  /// https://leetcode.com/problems/reveal-cards-in-increasing-order
  /// 
  /// Difficulty Medium
  /// Acceptance 83.0%
  /// 
  /// Array
  /// Queue
  /// Sorting
  /// Simulation
  /// </summary>
  public class Solution
  {
    public int[] DeckRevealedIncreasing(int[] deck)
    {
      var len = deck.Length;

      Array.Sort(deck);
      Array.Reverse(deck);

      var result = new Queue<int>(len);
      result.Enqueue(deck[0]);

      for (var i = 1; i < len; i++)
      {
        result.Enqueue(result.Dequeue());
        result.Enqueue(deck[i]);
      }

      deck = result.ToArray();
      Array.Reverse(deck);

      return deck;
    }
  }
}

File: Solutions.java/Simulation/0950. Reveal Cards In Increasing Order.java
/// <summary>
/// 950. Reveal Cards In Increasing Order
/// https://leetcode.com/problems/reveal-cards-in-increasing-order
/// 
/// Difficulty Medium
/// Acceptance 83.0%
/// 
/// Array
/// Queue
/// Sorting
/// Simulation
/// </summary>
class Solution
{
  public int[] deckRevealedIncreasing(int[] deck)
  {
    var len = deck.length;

    Arrays.sort(deck);
    reverse(deck);

    var result = new LinkedList<Integer>();
    result.add(deck[0]);

    for(var i = 1; i < len; i++)
    {
      result.add(result.poll());
      result.add(deck[i]);
    }

    for(var i = 0; i < len; i++)
      deck[i] = result.get(i);
      
    reverse(deck);

    return deck;
  }

  private void reverse(int[] data)
  {
    for(int i = 0; i < data.length / 2; i++)
    {
      var temp = data[i];
      data[i] = data[data.length - i - 1];
      data[data.length - i - 1] = temp;
    }
  }
}
